Fix: Excerpt limit is overruled (#48403)

tmanhollan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
REST_REQUEST constant was mistakenly quoted, which caused the condition to always evaluate to true. And the conditional logic should be AND, not OR.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Issue #48403 describes the problem. Excerpt length is overridden always, rather than just in the editor as intended.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Fixes the conditional logic so implementations of excerpt_length filter hook will only be overridden in the block editor.

## Testing Instructions
1. Implement the following hook:
```php
add_filter( 'excerpt_length', function( $excerpt_length ) {
	return 1000;
}, 100 );
```
2. Add or edit a post in the block editor.
3. Use the post-excerpt block.

### Testing Instructions for Keyboard
n/a

## Screenshots or screencast
n.a

tmanhollan committed:
Fix Excerpt limit is overruled (#48403)

REST_REQUEST constant was mistakenly quoted, which caused the condition to always evaluate to true. And the conditional logic should be AND, not OR.

skorasaurus commented:
Thanks for the quick pull request and initial report. 

This PR did fix the issue of the `excerpt_length` filter being ignored in my PHP templates ; however in the editor, the post excerpts block's content length is limited to the upper bound of the excerpt_length filter. For example, if you have excerpt_length filter set to 55 but in the block editor, set the block's excerpt to 100; the excerpt will only display 55 words in the editor... if you change block's excerpt to 25 in the editor; the block will correctly display 25 words.

tmanhollan commented:
Thank you, @skorasaurus, you're right.

I think there are two problems: primarily, the condition is not being evaluated in the correct context. It should be inside the callback; in the outer scope I think it's happening before `REST_REQUEST` is set. Additionally, it appears `is_admin()` isn't true for the rest request that returns the post(s). In fact, I don't think that even needs to be part of the condition, but I'm not sure if there's some other case where it might be true.

I'm making that revision now. But I don't think `REST_REQUEST` being true is sufficient on its own for knowing that it's being loaded in the editor. Is there more to consider there? I really don't know that much about this API; I just discovered the bug while trying to figure out why my excerpt length was being ignored.

tmanhollan committed:
Evaluate condition inside the callback

Outer scope happens before the constant is set.

tmanhollan committed:
Fix coding standards

skorasaurus commented:
Thanks; with [your latest commit](https://github.com/WordPress/gutenberg/pull/48598/commits/21367410f67674fd5ecfe9d0280c42c62d81daf4), the post_excerpt block was now working again within the editor although someone else had caught it and fixed it (https://github.com/WordPress/gutenberg/pull/48654) but they didn't wrap it up in a callback.  I haven't checked yet if their commit, already merged into trunk fixes all of these issues.

skorasaurus commented:
Query loop block: 


(using https://github.com/WordPress/gutenberg/pull/48654)
when:
excerpt_length hook is set to 50 
excerpt length block's attribute set to 85
what displays in editor: value of excerpt_length hook 
what displays in front end: value of excerpt_length hook 

excerpt_length hook is set to 50
excerpt length block's attribute set to 25 
what displays in editor: excerpt length block's attribute
what displays in front end: excerpt length block's attribute

-------------------


for #48598 as of https://github.com/WordPress/gutenberg/blob/21367410f67674fd5ecfe9d0280c42c62d81daf4/packages/block-library/src/post-excerpt/index.php

excerpt_length hook is set to 50 
excerpt length block's attribute set to 85
what displays in editor: excerpt length block's attribute
what displays in front end: value of excerpt_length hook 


excerpt_length hook is set to 50
excerpt length block's attribute set to 25 
what displays in editor: excerpt length block's attribute
what displays in front end: excerpt length block's attribute

tmanhollan commented:
I think I see the problem in the front end. It's not loaded via REST like it is in the editor, so the condition in the filter is not satisfied (the only reason it was originally is that the condition always returned true because of the quoted constant). I think we need another excerpt_length filter applied in `render_block_core_post_excerpt` right before `get_the_excerpt` is called to make sure the starting value isn't trimmed shorter than the max that can be set in the block's setting. I'll try that out now.

tmanhollan committed:
Filter the excerpt length for the front end output

The filter that is conditional upon REST_REQUEST does not apply when the block content is rendered in the front end. This added filter serves the same purpose when rendering the block for the front end.

tmanhollan commented:
Ok I think [this last revision](https://github.com/WordPress/gutenberg/pull/48598/commits/dc82d46dc137b337eaa9433a2c8ff9f3fe7758f3) makes the front end and editor consistent and excerpts that are not part of a block obey the value returned by the excerpt_length hook.

excerpt_length hook is set to 50
excerpt length block's attribute set to 85
displayed in the editor: value of attribute (85)
displayed in the front end: value of attribute (85)
regular excerpt on another page (not part of the post-excerpt block): value of of hook (50)

excerpt_length hook is set to 50
excerpt length block's attribute set to 25
displayed in the editor: value of attribute (25)
displayed in the front end: value of attribute (25)
regular excerpt on another page (not part of the post-excerpt block): value of of hook (50)

skorasaurus commented:
(For anyone new coming to this: even though the original issue #48403 is closed; this should be kept open to determine whether this is an improvement of https://github.com/WordPress/gutenberg/pull/48654; also see review at https://github.com/WordPress/gutenberg/pull/44964#pullrequestreview-1321446813 in https://github.com/WordPress/gutenberg/pull/44964

carolinan commented:
I am not able to do the code review, I am not confident I would get it right since I caused the bugs last time.
From the [comment](https://github.com/WordPress/gutenberg/pull/48598#issuecomment-1451053507) it does look like this is an improvement following the smaller fix, but the solution for the last filter is different, I can't determine which is the right solution.

Mamaduka commented:
Thanks for contributing, @tmanhollan!

It looks like the linked issue has been resolved. There's an alternative proposal to remove this filter hack, which appears to be a preferable resolution here. Ref: #69090.

I'm going to close this PR.
